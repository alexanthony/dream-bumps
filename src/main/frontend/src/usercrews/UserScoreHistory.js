import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import {Card, CardTitle, CardText} from 'material-ui/Card';
import { Line } from 'react-chartjs-2';

import { fetchUserCrewPriceHistory } from './actions';
import { getUserScoreHistory } from '../rootReducer';

const minPrice = scoreHistories => scoreHistories.reduce((prev, curr) => (prev < (curr.value + curr.cash) ? prev : (curr.value + curr.cash)), Number.POSITIVE_INFINITY);

const maxPrice = scoreHistories => scoreHistories.reduce((prev, curr) => (prev > (curr.value + curr.cash) ? prev : (curr.value + curr.cash)), 0);

/**
 * CrewPriceHistory
 */
class UserScoreHistory extends Component {
  componentWillMount() {
    if (this.props.userId) {
      this.props.fetchUserCrewPriceHistory(this.props.userId, this.props.sex);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.userId && nextProps.userId !== this.props.userId) {
      this.props.fetchUserCrewPriceHistory(nextProps.userId, nextProps.sex);
    }
  }

  render() {
    return (
      <Card style={this.props.style}>
        <CardTitle title="Score History" />
        <CardText>
          <Line
            data={{datasets: [{data: this.props.scoreHistory.map(history => ({x: history.dateTime, y: history.value + history.cash })),
                      fill: false,
                      steppedLine: true,}]}}
            options={{
                scales: {
                  xAxes: [{type:'time'}],
                  yAxes: [{
                    ticks: {
                      suggestedMin: minPrice(this.props.scoreHistory) - 10,
                      suggestedMax: maxPrice(this.props.scoreHistory) + 10
                  }}]
                },
                legend: {
                  display: false
                }
              }}
          />
        </CardText>
      </Card>
    );
  }
}

UserScoreHistory.propTypes = {
  scoreHistory: PropTypes.array,
  userId: PropTypes.number,
  fetchUserCrewPriceHistory: PropTypes.func,
  sex: PropTypes.string,
  style: PropTypes.object
};

export const mapStateToProps = (state, ownProps) => ({
  scoreHistory: getUserScoreHistory(state)(ownProps.userId, ownProps.sex) || []
});

export {UserScoreHistory};

export default connect(mapStateToProps, { fetchUserCrewPriceHistory })(UserScoreHistory);
