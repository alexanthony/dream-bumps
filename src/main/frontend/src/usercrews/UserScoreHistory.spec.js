import React from 'react';
import { shallow } from 'enzyme';
import { UserScoreHistory } from './UserScoreHistory';
import toJson from 'enzyme-to-json';

test('Renders properly', () => {
  const fetchUserCrewPriceHistory = jest.fn();
  const scoreHistories = [
    {cash: 100.12, value: 1027.8, dateTime: '2017-05-02T21:51:15.123Z'},
    {cash: 102.34, value: 1025.78, dateTime: '2017-05-02T21:53:15.123Z'},
    {cash: 99.23, value: 1023.45, dateTime: '2017-05-04T12:51:15.123Z'}
  ];
  const component = shallow(
    <UserScoreHistory
      userId={7}
      sex="male"
      style={{marginTop: '16px'}}
      fetchUserCrewPriceHistory={fetchUserCrewPriceHistory}
      scoreHistory={scoreHistories}
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
  // We don't actually mount, so these don't get called...
  // expect(fetchUsers).toHaveBeenCalled();
  // expect(fetchUserCrewRankings).toHaveBeenCalled()
});


// UserScoreHistory.propTypes = {
//   scoreHistory: PropTypes.array,
//   userId: PropTypes.number,
//   fetchUserCrewPriceHistory: PropTypes.func,
//   sex: PropTypes.string,
//   style: PropTypes.object
// };
