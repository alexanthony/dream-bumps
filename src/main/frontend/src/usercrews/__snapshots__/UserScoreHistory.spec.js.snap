exports[`test Renders properly 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "marginTop": "16px",
    }
  }>
  <CardTitle
    title="Score History" />
  <CardText>
    <Line
      data={
        Object {
          "datasets": Array [
            Object {
              "data": Array [
                Object {
                  "x": "2017-05-02T21:51:15.123Z",
                  "y": 1127.92,
                },
                Object {
                  "x": "2017-05-02T21:53:15.123Z",
                  "y": 1128.12,
                },
                Object {
                  "x": "2017-05-04T12:51:15.123Z",
                  "y": 1122.68,
                },
              ],
              "fill": false,
              "steppedLine": true,
            },
          ],
        }
      }
      options={
        Object {
          "legend": Object {
            "display": false,
          },
          "scales": Object {
            "xAxes": Array [
              Object {
                "type": "time",
              },
            ],
            "yAxes": Array [
              Object {
                "ticks": Object {
                  "suggestedMax": 1138.12,
                  "suggestedMin": 1112.68,
                },
              },
            ],
          },
        }
      } />
  </CardText>
</Card>
`;
